Readme.txt
FAM-N (Fast Amplitude Modulation - N pulses) is a composite pulse that
exploits the phenomenon of inversion/saturation of satellite transtion into the central
transition to enhance the efficiency of the conversion pulse in MQMAS
experiments, improving the overall sensitivity of the experiment.
This program determines the number of individual pulse and their
respective durations required for an optimal conversion under the
experimental conditions defined by the user.

See "Manual/quickstart.pdf" for instructions

- More information: 
Colaux, H.; Dawson, D. M. & Ashbrook, S. E.,
Efficient Ampli tude-Modulated Pulses for Triple-to Single-Quantum Coherence Conversion in MQMAS NMR,
The Journal of Physical Chemistry A, 2014

FAMN optimisation program requires SIMPSON to be installed in order to function. SIMPSON is
released under the conditions of the GNU General Public License.
SIMPSON can be obtained from <http://nmr.au.dk/software/simpson/>

- More information: 
Bak, M.; Rasmussen, J.; Nielsen, N. & others,
SIMPSON: a general simulation program for solid-state NMR spectroscopy
Journal of Magnetic Resonance, 2000, 147, 296

# -----------------------------------------------------------
FAMN optimisation program (C) 2013-2019 Henri Colaux
FAMN optimisation program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, version 3.

FAMN optimisation program is distributed in the hopeq that it will be useful,
but without any warranty; without even the implied warranty of
 merchantability or fitness for a particular purpose.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with FAMN optimisation program.  If not, see <http://www.gnu.org/licenses/>.

# -----------------------------------------------------------
Henri Colaux
e-mail: henri.colaux@normalesup.org

Changelog.txt
— Version 0.3.2a
	– Various bug fixes

– Version 0.3a:
    – The number of points now refers to a fixed nutation
    – Added the option to stop 'min_improvement' : Minumum improvement
	

famn_opt/old_debug/test_machin.py
# -*- coding: utf-8 -*-
"""
Console mode execution of the FAM-N optimisation program.
Fill in the prameters execute this script (console)
or press the run button/F5 (Spyder/pyCharm/VSstudio…)
"""

# Initialise the script
from FAMN_opt import FAMN_main
test_machin = FAMN_main()
test_machin.parameters['simpson_basename'] = test_machin
# ----------------------------------------------------------------------------
# Input parameters

 # Spin system input parameters
test_machin.parameters['nucleus'] = '23Na' # Nucleus, as written in Simpson
test_machin.parameters['proton_larmor_frequency'] = 600e6 # Proton Larmor frequency (Hz) 
test_machin.parameters['MAS'] = 12500 # Magic angle spinning rate (Hz) 
test_machin.parameters['RF'] = 100000 # Radio-frequency field ν1 (Hz)  
test_machin.parameters['CQ'] = 1.0e6 # Quadrupolar coupling constant CQ (Hz) 
test_machin.parameters['ηQ'] = 0 # Asymmetry parameter EtaQ 

# ----------------------------------------------------------------------------
# Transfer parameters

# To fill for a standard FAM-N optimisation
test_machin.parameters['transition'] = 3 # Transition of the FAM-N method.

# To fill in for complex FAM-N optimisation
test_machin.parameters['complex_optimisation'] = ''
test_machin.parameters['single_pulse_nutation_list'] = '', # If complex optimisation, string list of the single pulse nutation angles for each step

# ---------------------------------------------
# Advanced user parameters.

test_machin.parameters['min_improvement'] = 1.1 # Minimum improvement between two subsequent pulses
                                    # for the simulation to be allowed to add additionnal pulses
test_machin.parameters['max_nopulses'] = int(30) # Maximum number of pulses that the program is allowed to output.
test_machin.parameters['nop_360'] = 180 # Number of points par 360° nutation.
test_machin.parameters['min_time_increment'] = 1e-8 # Minimum time increment (seconds)
test_machin.parameters['maxdt_ratio'] = int(20) # Ratio between MAS period and maxdt in SIMPSON.
test_machin.parameters['crystfile'] = 'rep66' # Crystal file used for the SIMPSON optimisation
test_machin.parameters['no_γ'] = 4 # Number of γ-angles for the SIMPSON optimisation
test_machin.parameters['num_cores'] = int(0) # Number of cores on which the optimisation is programmed

# ----------------------------------------------------------------------------
# Execution of the code
test_machin.initialise()
test_machin.FAMN_loop()
test_machin.finalise()

# ----------------------------------------------------------------------------


famn_opt/old_debug/FAMN_opt_debug_final2.py
# -*- coding: utf-8 -*-
"""
Console mode execution of the FAM-N optimisation program.
Fil-in the prameters and press the run button/F5
@author: Henri
"""

import numpy
import subprocess
import FAMN_opt as fnf
import matplotlib.pyplot as plt

import copy
self = copy.deepcopy(famn.output['FAMNsteps'][1])
self.FAMN_main.parameters['no_SIMPSON_execution'] = False

#self = fnf.FAMN_main()
#self.loadstate('test_conc')


# Initialise some arrays
self.saved_density = []
self.saved_detect = []

self.temp['cur_name'] = self.finalise_filename(1)

print('Final simulation for pulse n°1')
# Do the simulations for the single pulse
            


# -----------------------------------------------------------------

# Write SIMPSON input file
self.write_simpson(self.FAMN_main.output['FAMNsteps']\
                   [self.opt['order_index']-1].data['final_density'], # Initial density matrix
                   self.opt['nop_CW'], # Number of points
                   self.FAMN_main.output['dim_density'], # Dimesion of the density matrix
                   ) # self.temp['pulse_phase'] # Pulse phase

# SIMPSON Execution
if not self.FAMN_main.parameters['no_SIMPSON_execution']:
    subprocess.call('simpson ' + './' + self.FAMN_main.parameters['simpson_tempfolder'] +\
          '/' + self.temp['cur_name'] + '.in')

# Read in the file
self.temp['simpson_out_full'] = fnf.readin_simpson_out(('./' + self.FAMN_main.parameters['simpson_tempfolder'] +\
      '/' + self.temp['cur_name'] + '.out'),
    self.opt['nop_CW'], self.FAMN_main.output['dim_density'],
    self.FAMN_main.output['size_density'],
    )

# Save the current density evolution
self.saved_density.append(self.temp['simpson_out_full'])

# Test figure
plt.figure(figsize = (12,6))
self.saved_detect.append(self.get_detect_element(self.saved_density[0],self.opt['nop_CW']))
plt.plot(1e6 * self.opt['time_increment'] * numpy.arange(0,self.opt['nop_CW']),self.saved_detect[0])

# -----------------------------------------------------------------
# Loop over the number of pulses
for truc in range(1,self.data['tnopules']):
    
    print('Final simulation for pulse n°' + str(truc+1))
    # self.temp['pulse_phase'] = not self.temp['pulse_phase'] # Inverse phase # Not used
    
    # Change filename
    self.temp['cur_name'] = self.finalise_filename(truc+1)
    
    # Write SIMPSON input file
    self.write_simpson(self.saved_density[truc-1][self.data['max_pos'][truc-1]+self.data['inv_pos'][truc-1]],
                       # Initial density matrix
                       self.data['pulse_dur'][truc], # Number of points
                       self.FAMN_main.output['dim_density'], # Dimesion of the density matrix
                       ) # self.temp['pulse_phase'] # Pulse phase
    
    # SIMPSON Execution
    if not self.FAMN_main.parameters['no_SIMPSON_execution']:
        subprocess.call('simpson ' + './' + self.FAMN_main.parameters['simpson_tempfolder'] +\
              '/' + self.temp['cur_name'] + '.in')
    
    # Read in the file
    self.temp['simpson_out_full'] = fnf.readin_simpson_out(('./' + self.FAMN_main.parameters['simpson_tempfolder'] +\
          '/' + self.temp['cur_name'] + '.out'),
        self.data['pulse_dur'][truc], self.FAMN_main.output['dim_density'],
        self.FAMN_main.output['size_density'],
        )
    
    # Save the current density evolution
    self.saved_density.append(self.temp['simpson_out_full'])
    self.saved_detect.append(self.get_detect_element(self.saved_density[truc],self.data['pulse_dur'][truc]))



    plt.plot(1e6 * self.opt['time_increment'] * numpy.arange(self.data['inv_pos_abs'][truc-1],
                          self.data['inv_pos_abs'][truc-1] + self.data['pulse_dur'][truc]),
                            self.saved_detect[truc])
    
    
    
    
    
    
    
    

famn_opt/old_debug/famn_opt_debug_finalise.py
# -*- coding: utf-8 -*-
"""
Console mode execution of the FAM-N optimisation program.
Fil-in the prameters and press the run button/F5
@author: Henri
"""

import subprocess
from famn_core import *
import matplotlib.pyplot as plt
import numpy
import copy

# self = copy.deepcopy(famn.output['FAMNsteps'][0])
self = copy.deepcopy(famn)

# ----------------------------------------------------------
# Final detect coherence
self.final_coherence = self.output['list_detect_density'][  self.output['no_FAMN_steps']-1  ]

# ----------------------------------------------------------
# Number of pulses
self.total_nopules = numpy.zeros(self.output['no_FAMN_steps'], dtype = 'int32')
for FAMNstepcount in range(self.output['no_FAMN_steps']): 
    self.total_nopules[FAMNstepcount] = int(self.output['FAMNsteps'][FAMNstepcount].data['tnopules'])
    if self.output['is_succesion'][FAMNstepcount]:
        self.total_nopules[FAMNstepcount-1] -= 1

# ----------------------------------------------------------
# Select the final coherence evolution
self.select_coherences = []
self.pulse_durations = []

for FAMNstepcount in range(self.output['no_FAMN_steps']):
    # Swipe all separated FAM-N of the optimisation
    if self.output['do_final_sim'][FAMNstepcount]:
        # 
        for FAMNpulses in range(self.output['FAMNsteps'][FAMNstepcount].data['tnopules']):
            self.select_coherences.append(get_all_detect_element(
                self.output['FAMNsteps'][FAMNstepcount].data['final_full_density'][FAMNpulses],
                self.final_coherence))
            self.pulse_durations.append(len(self.select_coherences[-1]))
    else:
            
        for FAMNpulses in range(self.output['FAMNsteps'][FAMNstepcount].data['tnopules']):
            self.select_coherences.append(get_all_detect_element(
                self.output['FAMNsteps'][FAMNstepcount].data['selected_full_density'][FAMNpulses],
                self.final_coherence))
            self.pulse_durations.append(len(self.select_coherences[-1]))


# -----------------------------------------------------
# Final signal improvements    
self.final_signal_improvement = numpy.real(self.select_coherences[-1][ self.output['FAMNsteps'][-1].data['max_pos'][-1] ])/\
                                numpy.real(self.select_coherences[0][ self.output['FAMNsteps'][0].data['max_pos'][0] ])

# -----------------------------------------------------
# Total number of points
self.total_nopoints = numpy.full((self.output['no_FAMN_steps']),int(0))  
for FAMNstepcount in range(self.output['no_FAMN_steps']):
    self.total_nopoints[FAMNstepcount] = numpy.sum(self.output['FAMNsteps'][FAMNstepcount].data['pulse_durations_nop'])

# Total duration: * self.output['FAMNsteps'][FAMNstepcount].opt['time_increment']

# -----------------------------------------------------
# Text fractions of the final pulse for output in Topspin and Delta
self.pulse_frac = []
for FAMNstepcount in range(self.output['no_FAMN_steps']): 
    # Pre-allocate
    temp = numpy.full( self.total_nopules[FAMNstepcount] , None )
    
    # Store in a temporary array
    for FAMNpulses in range(self.total_nopules[FAMNstepcount]):
        temp[FAMNpulses] = (str(self.output['FAMNsteps'][FAMNstepcount].data['pulse_durations_nop'][FAMNpulses]) +
                '/' + str(self.total_nopoints[FAMNstepcount]))
    # -----------------
    self.pulse_frac.append(temp)
    
# -----------------------------------------------------
# First absolute point of each pulse
self.initial_time = numpy.zeros(self.output['no_FAMN_steps'])
for FAMNstepcount in range(1,self.output['no_FAMN_steps']): 
    self.initial_time[FAMNstepcount] =\
    1.0e6 * self.output['FAMNsteps'][FAMNstepcount].opt['time_increment'] *\
    self.initial_time[FAMNstepcount-1] +\
    self.output['FAMNsteps'][FAMNstepcount].data['pulse_durations_nop'][-1]
    
    
# -----------------------------------------------------
# Pre-allocate figure
self.final_figure, self.final_axes = plt.subplots(1,1, figsize = (10,5))
self.final_axes.set_title('Final FAM-N pulse')
self.final_axes.set_xlabel('Time /μs')

counter = int(0)
for FAMNstepcount in range(self.output['no_FAMN_steps']):
    # Plot the first pulse
    self.final_axes.plot(self.initial_time +\
            1.0e6 *self.output['FAMNsteps'][FAMNstepcount].opt['time_increment'] *\
            numpy.arange(self.pulse_durations[counter]),
        self.select_coherences[counter])
    # Plot subsequent pulses
    counter =+ 1
    for FAMNpulses in range(self.output['FAMNsteps'][FAMNstepcount].data['tnopules']-1):
        # Plot curve
        self.final_axes.plot(self.initial_time + 1.0e6 * \
                self.output['FAMNsteps'][FAMNstepcount].opt['time_increment'] *numpy.arange(\
               self.output['FAMNsteps'][FAMNstepcount].data['inv_pos_abs'][FAMNpulses],
               self.output['FAMNsteps'][FAMNstepcount].data['inv_pos_abs'][FAMNpulses] +\
               self.pulse_durations[counter]),
               self.select_coherences[counter])
        # Trace point of phase inversion
        self.final_axes.axvline(x = self.initial_time[FAMNstepcount] +\
                        1.0e6 *self.output['FAMNsteps'][FAMNstepcount].opt['time_increment'] *\
                        self.output['FAMNsteps'][FAMNstepcount].data['inv_pos_abs'][FAMNpulses]        
                        , color = 'black')
        # -----
        counter =+ 1
# ---------------------------------------------------------------------
# Plot position of first maximum
self.final_axes.axvline(1.0e6 *self.output['FAMNsteps'][FAMNstepcount].opt['time_increment'] *\
                        self.output['FAMNsteps'][0].data['max_pos_abs'][0],
        color = 'red', linestyle = '--')
# Plot position of last maximum
self.final_axes.axvline(self.initial_time[-1] +\
                        1.0e6 *self.output['FAMNsteps'][FAMNstepcount].opt['time_increment'] *\
                        self.output['FAMNsteps'][-1].data['max_pos_abs'][-1],
        color = 'red', linestyle = '--')








famn_opt/old_debug/FAMN_opt_debug_test_alignement.py
# -*- coding: utf-8 -*-
"""
Console mode execution of the FAM-N optimisation program.
Fil-in the prameters and press the run button/F5
@author: Henri
"""

import numpy
import subprocess
import FAMN_opt as fnf
import matplotlib.pyplot as plt

import copy
self = copy.deepcopy(famn.output['FAMNsteps'][0])
self.FAMN_main.parameters['no_SIMPSON_execution'] = False

#self = fnf.FAMN_main()
#self.loadstate('test_conc')


# Initialise some arrays
self.saved_density = []
self.saved_detect = []

self.temp['cur_name'] = 'debug_test1'



# -----------------------------------------------------------------

# Write SIMPSON input file
self.write_simpson(self.FAMN_main.output['list_initial_density'][self.opt['order_index']], # Initial density matrix
                   self.opt['nop_CW'], # Number of points
                   self.FAMN_main.output['dim_density'], # Dimesion of the density matrix
                   ) # self.temp['pulse_phase'] # Pulse phase

# SIMPSON Execution
if not self.FAMN_main.parameters['no_SIMPSON_execution']:
    subprocess.call('simpson ' + './' + self.FAMN_main.parameters['simpson_tempfolder'] +\
          '/' + self.temp['cur_name'] + '.in')

# Read in the file
self.simpson_out_full1 = fnf.readin_simpson_out(('./' + self.FAMN_main.parameters['simpson_tempfolder'] +\
      '/' + self.temp['cur_name'] + '.out'),
    self.opt['nop_CW'], self.FAMN_main.output['dim_density'],
    self.FAMN_main.output['size_density'],
    )


self.simpson_out_full1 = numpy.concatenate((
        [numpy.conj(self.FAMN_main.output['list_initial_density'][self.opt['order_index']])],
         self.simpson_out_full1))

# Get only the detect element from this numpy array
self.detect1 = numpy.real(self.get_detect_element(\
         self.simpson_out_full1,self.opt['nop_CW']+1))



plt.plot(self.detect1)



# -----------------------------------------------------------------

self.temp['cur_name'] = 'debug_test2'

nop = 5

# Write SIMPSON input file
self.write_simpson(numpy.conj(self.simpson_out_full1[nop]), # Initial density matrix
                   self.opt['nop_CW'], # Number of points
                   self.FAMN_main.output['dim_density'], # Dimesion of the density matrix
                   ) # self.temp['pulse_phase'] # Pulse phase

# SIMPSON Execution
if not self.FAMN_main.parameters['no_SIMPSON_execution']:
    subprocess.call('simpson ' + './' + self.FAMN_main.parameters['simpson_tempfolder'] +\
          '/' + self.temp['cur_name'] + '.in')




# Read in the file
self.simpson_out_full2 = fnf.readin_simpson_out(('./' + self.FAMN_main.parameters['simpson_tempfolder'] +\
      '/' + self.temp['cur_name'] + '.out'),
    self.opt['nop_CW'], self.FAMN_main.output['dim_density'],
    self.FAMN_main.output['size_density'],
    )


self.simpson_out_full2 = numpy.concatenate((
        [numpy.conj(self.simpson_out_full1[nop])],
         self.simpson_out_full2 ))

self.detect2 = self.get_detect_element(self.simpson_out_full2,self.opt['nop_CW']+1)



plt.plot(numpy.arange(nop,nop+self.opt['nop_CW']+1),self.detect2)
    
    
    

console.py
# -*- coding: utf-8 -*-
"""
Console mode execution of the FAM-N optimisation program.
Fil-in the prameters execute this script (console)
or press the run button/F5 (Spyder/pyCharm/VSstudio …)
@author: Henri Colaux
"""

# ----------------------------------------------------------------------------
# Create a new class instance
from opt import FAMN_main
famn = FAMN_main()
famn.param['simpson_basename'] = 'famn'

# ----------------------------------------------------------------------------
# Input parameters

famn.param['nucleus'] = '23Na' # Nucleus, as written in Simpson
famn.param['proton_larmor_frequency'] = 600e6 # Proton Larmor frequency (Hz) 
famn.param['MAS'] = 12500 # Magic angle spinning rate (Hz) 
famn.param['RF'] = 100000 # Radio-frequency field ν1 (Hz)  
famn.param['CQ'] = 1e6 # Quadrupolar coupling constant CQ (Hz) 
famn.param['ηQ'] = 0 # Asymmetry parameter ηQ 

famn.param['export_filepath'] = '', # (Optionnal) Path where all output files are exported.
# ----------------------------------------------------------------------------
# Transfer parameters

# To fill for a standard FAM-N optimisation
famn.param['transition'] = 3 # Transition of the FAM-N method.

# To fill in for complex FAM-N optimisation
famn.param['complex_optimisation'] = '' # '5>3;3>1'
famn.param['single_pulse_nutation_list'] = ''  # '120,75' # If complex optimisation, string list of the single pulse nutation angles for each step

# ---------------------------------------------
# Crystal file
famn.param['no_cryst'] = 66 # Number of crystallites
famn.param['no_γ'] = 4 # Number of γ-angles for the SIMPSON optimisation

# ---------------------------------------------
# Advanced user parameters.
famn.param['min_improvement'] = 1.1 # Minimum improvement between two subsequent pulses
                                    # for the simulation to be allowed to add additionnal pulses
famn.param['max_nopulses'] = int(30) # Maximum number of pulses that the program is allowed to output.
famn.param['nop_360'] = 180 # Number of points par 360° nutation.
famn.param['min_time_increment'] = 1e-8 # Minimum time increment (seconds)
famn.param['maxdt_ratio'] = int(20) # Ratio between MAS period and maxdt in SIMPSON.
famn.param['num_cores'] = int(0) # Number of cores on which the optimisation is programmed

# ---------------------------------------------
# Debug
famn.param['no_SIMPSON_execution'] = True
famn.param['debug_figure'] = True
famn.param['do_every_cryst'] = True
famn.param['keep_simpson_files'] = True

# ----------------------------------------------------------------------------
# Execution of the code
famn.initialise()
famn.FAMN_loop()
famn.finalise()
famn.screen_report()


## Export Bruker
#famn.export_topspin_MQMAS('bruker_MQMAS')
#famn.export_topspin_pulselist('bruker_pulselist')
#famn.export_topspin_MQF('bruker_MQF')
#famn.export_topspin_MQCPMGMAS('bruker_MQCPMGMAS')
#
## Export Jeol
#famn.export_delta_MQMAS('jeol_MQMAS')
#famn.export_delta_MQF('jeol_MQF')
#famn.export_delta_pulselist('jeol_pulselist')

# ----------------------------------------------------------------------------